File: exercism-solutions/julia/run-length-encoding/run-length-encoding.jl

mutable struct Counter
    value
    count::Int

    Counter(value) = new(value, 1)
end


function str(counter::Counter)
    (counter.count > 1 ? string(counter.count) : "") * counter.value
end


function encode(s)
    if s == ""
        return ""
    end

    encoded = [Counter(s[1])]
    for ch in s[2:end]
        if ch == encoded[end].value
            encoded[end].count += 1
        else
            push!(encoded, Counter(ch))
        end
    end
    return join(map(str, encoded))
end


function decode(s)
    out = ""
    n = 0
    for ch in s
        maybe_number = Int(ch) - 48
        if 0 <= maybe_number <= 9
            n *= 10
            n += maybe_number
        else
            out *= repeat(ch, max(n, 1))
            n = 0
        end
    end
    return out
end

File: exercism-solutions/clojure/run-length-encoding/src/run_length_encoding.clj
(ns run-length-encoding)

(defn rle [lst]
  "run-length-encoding for list of characters"
  ((fn [inp out]
     (if (empty? inp)
       (reverse out)
       (let [n (first (first out))
             x (second (first out))]
         (recur
          (next inp)
          (if (= (first inp) x)
            (conj (next out) (list (inc n) x))
            (conj out (list 1 (first inp))))))))
   lst '()))

(defn run-length-encode
  "encodes a string with run-length-encoding"
  [plain-text]
  (clojure.string/join
   (map
    (fn [x]
      ; special case for skipping 1's
      (if (= (first x) 1)
        (second x)
        (clojure.string/join x)))
    (rle (seq plain-text)))))

(defn split-symbols [lst]
  "split string to integers and characters"
  ((fn [inp out]
     (if (empty? inp)
       (reverse out)
       (let [x (first inp)]
         (if (Character/isDigit x)
           (if (int? (first out))
             (recur
              (next inp)
              (conj
               (next out)
               ; collect digits as a single number
               (+ (* (first out) 10)
                  (Character/digit x 10))))
             (recur
              (next inp)
              (conj out (Character/digit x 10))))
           (recur
            (next inp)
            (conj out x))))))
   lst '()))

(defn run-length-decode
  "decodes a run-length-encoded string"
  [cipher-text]
  ((fn [inp out]
     (if (empty? inp)
       out
       (let [h (first inp)
             t (next inp)]
         (if (char? h)
           (recur t (str out h))
           (recur
            (next t)
            (str out (clojure.string/join
                      "" (repeat h (first t)))))))))
   (split-symbols cipher-text) ""))
